common/lock.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2021 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+#include "includes.h"
+
+static tdnflock tdnflock_new(const char *lock_path, const char *descr)
+{
+	mode_t oldmask;
+	uint32_t dwErr = 0;
+	tdnflock lock = NULL;
+
+	if (IsNullOrEmptyString(lock_path) || IsNullOrEmptyString(descr))
+	{
+		goto error;
+	}
+
+	dwErr = TDNFAllocateMemory(1, sizeof(*lock), (void *)&lock);
+	if (dwErr)
+	{
+		goto error;
+	}
+
+	oldmask = umask(022);
+	lock->fd = open(lock_path, O_RDWR|O_CREAT, 0644);
+	(void) umask(oldmask);
+
+	if (lock->fd == -1)
+	{
+		if (errno == EACCES)
+		{
+			lock->fd = open(lock_path, O_RDONLY);
+		}
+
+		if (lock->fd == -1)
+		{
+			TDNF_SAFE_FREE_MEMORY(lock);
+			lock = NULL;
+		}
+		else
+		{
+			lock->openmode = TDNFLOCK_READ;
+		}
+	}
+	else
+	{
+		lock->openmode = TDNFLOCK_WRITE | TDNFLOCK_READ;
+	}
+
+	if (lock)
+	{
+		dwErr = TDNFAllocateString(lock_path, &lock->path);
+		BAIL_ON_TDNF_ERROR(dwErr);
+		dwErr = TDNFAllocateString(descr, &lock->descr);
+		BAIL_ON_TDNF_ERROR(dwErr);
+		lock->fdrefs = 1;
+	}
+
+	return lock;
+
+error:
+	if (lock)
+	{
+		TDNF_SAFE_FREE_MEMORY(lock->path);
+		TDNF_SAFE_FREE_MEMORY(lock->descr);
+		TDNF_SAFE_FREE_MEMORY(lock);
+	}
+
+	return lock;
+}
+
+static void tdnflock_free(tdnflock lock)
+{
+	if (lock && --lock->fdrefs == 0)
+	{
+		TDNF_SAFE_FREE_MEMORY(lock->path);
+		TDNF_SAFE_FREE_MEMORY(lock->descr);
+		(void) close(lock->fd);
```

Comment:
Check for valid `fd` with `if (lock-fd != -1)`, so you can use it in the error case in `tdnflock_new`.